src/toast/tod/crosstalk.py

```diff
@@ -0,0 +1,587 @@
+    instead of initiating directly.
+    """
+
+    def __init__(
+        self,
+        n_crosstalk_matrices: 'int',
+        crosstalk_matrices: 'List[SimpleCrosstalkMatrix]',
+        *,
+        name: 'str' = "crosstalk",
+        debug: 'bool' = False,
+    ):
+        self.n_crosstalk_matrices = n_crosstalk_matrices
+        self.crosstalk_matrices = crosstalk_matrices
+        self.name = name
+        self.debug = debug
+
+    def _get_crosstalk_matrix(self, i: 'int') -> 'SimpleCrosstalkMatrix':
+        """Get the i-th crosstalk matrix, used this with MPI only.
+        """
+        debug = self.debug
+
+        rank_owner = i % PROCS
+        # index of the i-th matrix in the local rank
+        idx = i // PROCS
+
+        if RANK == rank_owner:
+            crosstalk_matrix = self.crosstalk_matrices[idx]
+
+            names = crosstalk_matrix.names
+            data = crosstalk_matrix.data
+
+            # cast to int for boardcasting
+            names_int = names.view(np.uint8)
+            # the data from HDF5 is already float64
+            # this is needed for comm.Bcast below
+            data = data.view(np.float64)
+
+        # prepare lengths for creating arrays
+            lengths = np.array([names.size, names.dtype.itemsize], dtype=np.int64)
+        else:
+            lengths = np.empty(2, dtype=np.int64)
+        world_comm.Bcast(lengths, root=rank_owner)
+        if debug:
+            logger.debug(f'crosstalk: Rank {RANK} receives lengths {lengths}')
+
+        # broadcast arrays
+        if RANK != rank_owner:
+            n = lengths[0]
+            name_len = lengths[1]
+            names_int = np.empty(n * name_len, dtype=np.uint8)
+            names = names_int.view(f'S{name_len}')
+            data = np.empty((n, n), dtype=np.float64)
+        world_comm.Bcast(names_int, root=rank_owner)
+        if debug:
+            logger.debug(f'crosstalk: Rank {RANK} receives names {names}')
+        world_comm.Bcast(data, root=rank_owner)
+        if debug:
+            logger.debug(f'crosstalk: Rank {RANK} receives data {data}')
+
+        return (
+            crosstalk_matrix
+        ) if RANK == rank_owner else (
+            SimpleCrosstalkMatrix(names, data)
+        )
+
+    @staticmethod
+    def _read_serial(
+        paths: 'List[Path]',
+        *,
+        debug: 'bool' = False,
+    ) -> 'List[SimpleCrosstalkMatrix]':
+        """Read crosstalk matri(x|ces) from HDF5 file(s) serially."""
+        return [SimpleCrosstalkMatrix.load(path, debug=debug) for path in paths]
+
+    @staticmethod
+    def _read_mpi(
+        paths: 'List[Path]',
+        *,
+        debug: 'bool' = False,
+    ) -> 'List[SimpleCrosstalkMatrix]':
+        """Read crosstalk matri(x|ces) from HDF5 file(s) with MPI.
+
+        This holds only those matrices owned by a rank
+        dictate by the condition `i % PROCS == RANK`
+        """
+        N = len(paths)
+        path_idxs_per_rank = range(RANK, N, PROCS)
```

Comment:
Here is another function where the communicator should be passed in as an argument

Reply:
See https://github.com/hpc4cmb/toast/pull/380#discussion_r663285530 & https://github.com/hpc4cmb/toast/pull/380#discussion_r664973869

(But I will move the global var local to SimpleCrosstalkMatrix instead.)